src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -483,16 +508,12 @@ internal ODataResourceMetadataContextWithModel(ODataResourceBase resource, IODat
                         IEdmEntityType entityType = this.actualResourceType as IEdmEntityType;
                         if (entityType != null)
                         {
-                            this.keyProperties = keyProperties = GetPropertyValues(entityType.Key(), resource, entityType, /*isKeyProperty*/ true, this.requiresId).ToArray();
-
-                            if (this.requiresId)
-                            {
-                                ValidateEntityTypeHasKeyProperties(this.keyProperties, this.ActualResourceTypeName);
-                            }
+                            this.keyProperties = GetPropertyValues(entityType.Key(), resource, entityType, this.requiresId).ToArray();
                         }
-                        else
+
+                        if (!ValidateEntityTypeHasKeyProperties(this.keyProperties, this.ActualResourceTypeName, this.requiresId))
                         {
-                            this.keyProperties = Enumerable.Empty<KeyValuePair<string, object>>().ToArray();
+                            return Enumerable.Empty<KeyValuePair<string, object>>().ToArray();

```

Comment:
Probably not worth it as part of this change, but this will allocate a new array every time we return, even though all 0-length arrays are the same (we actually get warnings from the compiler for this). We could avoid unnecessary allocations and garbage collection with this small class:

```
public static class Array
{
  public static T[] Empty<T>()
  {
    return Internal<T>.Empty;
  }
        
  private static class Internal<T>
  {
    public static T[] Empty = new T[0];
  }
}
```

New versions of .NET already have these singletons, but for back-compat we could write this one ourselves.

Reply:
Yes; I tried to use Array.Empty<T>(), but discovered it wasn't supported in .net45, so left it as it was. I agree we should go through and optimize everywhere we return empty arrays.